File: medium/graph/redundant_connection.py
import collections
from typing import List, Dict


class Solution:
    # noinspection PyMethodMayBeStatic
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:

        adjacency_list: Dict[int, List[int]] = {}
        queue = collections.deque()
        for edge in edges:
            if edge[0] in adjacency_list:
                adjacency_list[edge[0]].append(edge[1])
            else:
                adjacency_list[edge[0]] = [edge[1]]
            if edge[1] in adjacency_list:
                adjacency_list[edge[1]].append(edge[0])
            else:
                adjacency_list[edge[1]] = [edge[0]]

        for key, value in adjacency_list.items():
            if len(value) == 1:
                queue.append(key)

        while len(queue) > 0:
            node = queue.pop()
            linked_edges = adjacency_list.pop(node)
            adjacency_list[linked_edges[0]].remove(node)
            if len(adjacency_list[linked_edges[0]]) == 1:
                queue.append(linked_edges[0])

        for i in range(len(edges)-1, -1, -1):
            if edges[i][0] in adjacency_list and edges[i][1] in adjacency_list:
                return edges[i]

File: src/main/java/leet/solutions/RedundantConnection.java
package leet.solutions;

/*
https://leetcode.com/problems/redundant-connection/
Leetcode#684
 */

import java.util.*;

public class RedundantConnection {

    public int[] solve(int[][] edges) {
        Solution solution = new Solution();
        return solution.findRedundantConnection(edges);
    }


    @SuppressWarnings("InnerClassMayBeStatic")
    class Solution {
        public int[] findRedundantConnection(int[][] edges) {

            Map<Integer, Set<Integer>> adjacencyList = new HashMap<>();
            for (int[] edge : edges) {
                if (adjacencyList.containsKey(edge[0])) {
                    adjacencyList.get(edge[0]).add(edge[1]);
                } else {
                    adjacencyList.put(edge[0], new HashSet<>() {
                        {
                            add(edge[1]);
                        }
                    });
                }

                if (adjacencyList.containsKey(edge[1])) {
                    adjacencyList.get(edge[1]).add(edge[0]);
                } else {
                    adjacencyList.put(edge[1], new HashSet<>() {
                        {
                            add(edge[0]);
                        }
                    });
                }

            }

            LinkedList<Integer> evaluatorQueue = new LinkedList<>();
            for (Map.Entry<Integer, Set<Integer>> node : adjacencyList.entrySet()) {
                if (node.getValue().size() == 1) {
                    evaluatorQueue.addLast(node.getKey());
                }
            }

            while (evaluatorQueue.size() != 0) {
                Integer node = evaluatorQueue.pollFirst();
                Integer linkedNode = adjacencyList.get(node).stream().findFirst().orElse(-1);
                adjacencyList.get(linkedNode).remove(node);
                adjacencyList.remove(node);
                if (adjacencyList.get(linkedNode).size() == 1) {
                    evaluatorQueue.addLast(linkedNode);
                }
            }

            for (int i=edges.length-1; i >=0; i--)
            {
                if (adjacencyList.containsKey(edges[i][0]) && adjacencyList.containsKey(edges[i][1]))
                {
                    return edges[i];
                }
            }

            return new int[]{0,0};

        }
    }

}
